More leaf operators and aggregation in morsel runtime

pontusmelke opened the pull request:
- Use cursors directly
- Support label scan
- Support index seek
- Don't force operators to be handled by single thread
- Some aggregation operators: `count`, `min`, `max`, `avg`, `collect`

pontusmelke committed:
Use cursor directly in AllNodeScanOperator

pontusmelke committed:
Use cursor directly in ExpandAllOperator

pontusmelke committed:
Support label scans in morsel runtime

pontusmelke committed:
Support index seeks in morsel runtime

pontusmelke committed:
Do not loop endlessly

pontusmelke committed:
Fix warnings in morsel runtime

pontusmelke committed:
Remove restriction to single thread

pontusmelke committed:
Reduce morsel size

Using a morsel size of 100000 degrades performance at least on my machine,
and there are no other indications why 100000 is the _correct_ choice.

pontusmelke committed:
Support count with no grouping in morsel runtime

In doing so we also have created some of the foundations to be used
by other aggregating expressions.

pontusmelke committed:
Support count with no grouping in morsel runtime

In doing so we also have created some of the foundations to be used
by other aggregating expressions.

pontusmelke committed:
Unit tests for mappers and reducers

pontusmelke committed:
Add some acceptance tests

pontusmelke committed:
Add addition and subtraction for NumberValues

Implement overflow safe addition and subtraction on the supported
number values. Result of all operations are either a `LongValue` or
if the result doesn't fit in a long result is widened to a `DoubleValue`.

pontusmelke committed:
Support mutiplication of number values

Implement overflow aware multiplication on NumberValue

pontusmelke committed:
Make all math methods public

pontusmelke committed:
Support avg in morsel runtime

pontusmelke committed:
Support max and min in morsel runtime

pontusmelke committed:
Support count(*) in morsel runtime

pontusmelke committed:
Support collect in morsel runtime

pontusmelke committed:
Properly handle empty average

pontusmelke committed:
Some minor clean up

fickludd commented on community/values/src/main/java/org/neo4j/values/utils/ValueMath.java:
```diff
@@ -0,0 +1,104 @@
+        }
+        catch ( ArithmeticException e )
+        {
+            return Values.doubleValue( (double) a + (double) b );
```
Even though it shouldn't be too frequent, is there anyway to do this without using exceptions as a code path?

fickludd reviewed (changes requested):
I merely graced over the morsel stuff, so I cannot really determine the quality of that. I did look properly at the Values maths methods though, and those look pretty good to me. I'm wondering though if it wouldn't be enough to have specialized operators (`plus`, `minus`, `times`, `etc`) only for the `long` and `double` primitives? As you only implemented maths for those anyway, I don't really see any benefit in having the smaller primitives.

pontusmelke committed:
Check instead of relying on exceptions

The previous implementation was betting on that overflow are very rare,
however from benchmarking the drop of performance was steeper than expected.
For no overflow it is faster however at 1 percent overflow the performance
drops to being 26 times slower.

pontusmelke commented:
So I was going to explain how you're wrong and that using the intrinsic `addExact` et al would be so much faster for the normal case that it is totally worth taking the extra hit in the rare cases that we get an overflow. However actual numbers seems to disagree with that hypothesis - and the performance drop was steeper than I expected (and the payoff was not what I would have hoped for).

Running benchmarks with no overflow and 1% overflow

```
ExactMath.check       0 1544.297 ±  52.957   ops/ms
ExactMath.check       1 1214.406 ± 106.689  ops/ms
ExactMath.intrinsic   0 1526.825 ±  75.823  ops/ms
ExactMath.intrinsic   1   86.094 ±   8.380  ops/ms
```
 So yes you're right.

For your other question we still need to be able to handle `byte`, `short`, `int`, and `float` since they might be coming from parameters, e.g.

```
MATCH (n) RETURN avg(n.prop)
```

The plan is that this should replace what is now done in `OverFlowAwareSum`,  `Add`, `TypeSafeMathSupport` and probably a couple of other places.

fickludd commented:
Wow, I'm impressed that you benchmarked it, I was just guessing =).

Ok, let's keep the all the specializations as you already coded them. We'll see later if they're needed or not.

fickludd commented:
@pontusmelke I think this looks good now, but it needs a little love.

pontusmelke committed:
Remove byte, short, int, float operations

Only keep long and double since we are always promoting everything
anyway.

pontusmelke committed:
Switch to use strict math by default

pontusmelke committed:
Use overflow safe math in AvgOperatorExpression

pontusmelke committed:
Remove distinct

pontusmelke commented:
@fickludd I gave it some more love. I realized that having the widening behavior by default is wrong and not how cypher works today. Currently we use strict math by default and only use the widening to double in `AVG`, so I switched to that so that we can replace `TypeSafeMathSupport` with these method (in a separate PR).  If we want the automatic widening on for all additions I think we need a `BigIntegerValue` or something which can be backed by a `BigInteger` so that we can guarantee that `a+b` is never lossy for integers.

pontusmelke committed:
Don't confuse + and -

pontusmelke committed:
Checkstyle

systay commented on enterprise/cypher/morsel-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/vectorized/PipelineBuilder.scala:
```diff
@@ -102,6 +103,26 @@ class PipelineBuilder(slotConfigurations: SlotConfigurations, converters: Expres
+              val currentSlot = slots.get(key).get
+              //we need to make room for storing aggregation value in
+              //source slot
+              source.slots.newReference(key, currentSlot.nullable, currentSlot.typ)
```
Hmmm... If I understand this correctly, what we'll do is to use the input morsel first to calculate the aggregate values, and then we write those values back into the same morsel. Could we not instead break the pipeline here? From what I can see, the result will be that the input-morsel will be wider than necessary (since the aggregation values cannot be written until after all the input has been consumed), and for the output, we only use the last ref-slots of the first row.

It should work but seems inefficient.

systay commented on enterprise/cypher/morsel-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/vectorized/expressions/MinOrMaxOperatorExpression.scala:
```diff
@@ -0,0 +1,104 @@
+
+
+
+
```
This file should go on a diet.

systay commented on enterprise/cypher/morsel-runtime/src/test/scala/org/neo4j/cypher/internal/runtime/vectorized/operators/AggregationMapperOperatorTest.scala:
```diff
@@ -0,0 +1,175 @@
+    val aggregation = new AggregationMapperOperator(slots,
+                                                    Array(AggregationOffsets(2, 2, DummyEvenNodeIdAggregation(0))),
+                                                    Array(GroupingOffsets(groupSlot1, groupSlot1,
+                                                                          new DummyExpression(stringValue("A"), stringValue("B"))),
```
The line width here is itching. Could we indent things differently maybe?

systay commented on enterprise/cypher/morsel-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/vectorized/PipelineBuilder.scala:
```diff
@@ -88,6 +103,45 @@ class PipelineBuilder(slotConfigurations: SlotConfigurations, converters: Expres
           source = source.addOperator(preSorting)
           new MergeSortOperator(ordering, slots)
 
+        case plans.Aggregation(_, groupingExpressions, aggregationExpression) if groupingExpressions.isEmpty =>
```
Hmmm... If I understand this correctly, what we'll do is to use the input morsel first to calculate the aggregate values, and then we write those values back into the same morsel. Could we not instead break the pipeline here? From what I can see, the result will be that the input-morsel will be wider than necessary (since the aggregation values cannot be written until after all the input has been consumed), and for the output, we only use the last ref-slots of the first row.

It should work but seems inefficient.

(Same thing for both these cases)

fickludd replied:
@pontusmelke @systay If you two agree on this, please make a card to remember that optimization. I will merge this PR now before it gets stale.

pontusmelke replied:
👍
